src/host-config/proxy.ts

```diff
@@ -0,0 +1,275 @@
+		const rawRedsocksBlockMatch = rawConf.match(blockRegexFor('redsocks'));
+		// If match length shorter than 2, no group was captured, indicating malformed config
+		if (!rawRedsocksBlockMatch || rawRedsocksBlockMatch.length < 2) {
+			log.warn('Invalid redsocks block in redsocks.conf');
+			return conf;
+		}
+		const rawRedsocksBlock = RedsocksConf.parseBlock(
+			rawRedsocksBlockMatch[1],
+			disallowedProxyFields,
+		);
+		const maybeProxyConfig = ProxyConfig.decode(rawRedsocksBlock);
+		if (isRight(maybeProxyConfig)) {
+			conf.redsocks = {
+				...maybeProxyConfig.right,
+			};
+		} else {
+			log.warn(Reporter.report(maybeProxyConfig).join('\n'));
+			// Allow the invalid values for known fields through for
+			// backwards compatibility, while excluding unknown fields.
+			const maybeInvalidProxyConfig: Record<string, string> = {};
+			for (const key in rawRedsocksBlock) {
+				if (isAllowedProxyField(key)) {
+					maybeInvalidProxyConfig[key] = rawRedsocksBlock[key];
+				}
+			}
+
+			if (Object.keys(maybeInvalidProxyConfig).length) {
+				conf.redsocks = maybeInvalidProxyConfig as unknown as ProxyConfig;
+			}
+		}
+
+		return conf;
+	}
+
+	private static stringifyBlock(
+		label: string,
+		block: Record<string, any>,
+	): string {
+		const lines = Object.entries(block).map(([key, value]) => {
+			if (isAuthField(key)) {
+				// Add double quotes around login and password fields
+				value = `${value.startsWith('"') ? '' : '"'}${value}${value.endsWith('"') ? '' : '"'}`;
+			}
+			return `\t${key} = ${value};`;
+		});
+		return `${label} {\n${lines.join('\n')}\n}\n`;
+	}
+
+	/**
+	 * Given the raw contents of a block redsocks.conf file,
+	 * extract to a key-value object.
+	 */
+	private static parseBlock(
+		rawBlockConf: string,
+		unsupportedKeys: string[],
+	): Record<string, string> {
+		const parsedBlock: Record<string, string> = {};
+
+		// Split by newline and optional semicolon
+		for (const line of rawBlockConf.split(/;?\n/)) {
+			if (!line.trim().length) {
+				continue;
+			}
+			let [key, value] = line.split(/ *?= *?/).map((s) => s.trim());
+			// Don't parse unsupported keys
+			if (key && unsupportedKeys.some((k) => key.match(k))) {
+				continue;
+			}
+			if (key && value) {
+				if (isAuthField(key)) {
+					// Remove double quotes from login and password fields for readability
+					value = value.replace(/"/g, '');
+				}
+				parsedBlock[key] = value;
+			} else {
+				// Skip malformed lines
+				log.warn(
+					`Ignoring malformed redsocks.conf line ${isAuthField(key) ? `"${key}"` : `"${line.trim()}"`} due to missing key, value, or "="`,
+				);
+			}
+		}
+
+		return parsedBlock;
+	}
+}
+
+export async function readProxy(): Promise<HostProxyConfig | undefined> {
+	// Get and parse redsocks.conf
+	let rawConf: string | undefined;
+	try {
+		rawConf = await readFromBoot(redsocksConfPath, 'utf-8');
+	} catch (e: unknown) {
+		if (!isENOENT(e)) {
+			throw e;
+		}
+		return undefined;
```

Comment:
Why return undefined here instead of an empty object `{}` which is also a valid host proxy config?

Reply:
I think this is more a syntactic preference and makes the syntax in your [other](https://github.com/balena-os/balena-supervisor/pull/2329#discussion_r1657701558) comment better as well, but makes other parts of the host-config module more complicated as a result. For example, with `hostConfig.get`:
```
export async function get(): Promise<HostConfiguration> {
	const proxy = await readProxy();
	return {
		network: {
			hostname: await readHostname(),
			// Only return proxy if readProxy is not undefined
			...(proxy && { proxy }),
		},
	};
}
```
The return signature expects `HostConfiguration` which is
```
{
  network: t.partial({
    proxy: HostProxyConfig,
    hostname: t.string
  })
}
```
which means that if `readProxy` resolves `HostProxyConfig | Record<string, never>` instead of `HostProxyConfig | undefined`, even `...(Object.keys(proxy).length && { proxy })` is not enough for TypeScript not to complain as there is insufficient overlap between `undefined` and `Record<string, never>`, proxy must be explicitly cast to `HostConfiguration['network']['proxy']` for the return type to be satisfied. I don't think it's worth it for an internal method.

Reply:
I was wrong before, `{}` is not a valid `HostProxyConfig` so returning `undefined` is right. I guess that's what you meant by returning `Record<string, never>`